### PR TITLE
Feature/goes

### PR DESCRIPTION
**Pull request by zebengberg**
### Features

- New experimental `GOES` interface for downloading and visualizing GOES-16 satellite imagery.

## Tests

- [x] QC test passes locally (`make test`)
- [x] CI tests pass

## Reviewer

@mlshapiro 
@roger-teoh 
@trdean1 

No new unit tests, however, most of the new code **is doctested**. I'll also create a notebook showcasing usage.

### CODE REVIEW

**Comment by zebengberg** [comment]
There's a breaking change in xarray 2023.09 that is causing a unit test to fail (independent of the changes here). I'll fix this as part of this PR.

**Comment by trdean1** [comment]
This is super awesome! Excited that this will be part of the main repository.  See my minor issues above but otherwise great work!

File: pycontrails/datalib/goes.py [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+    level = "L2"  # Level 2
+    product_name = "CMIP"  # Cloud and Moisture Imagery
+    product = f"{sensor}-{level}-{product_name}{region.name[0]}"
+
+    bucket = bucket.removeprefix("gs://")
+
+    path_prefix = f"gs://{bucket}/{product}/{year}/{yday}/{hour}/"
+
+    # https://www.goes-r.gov/users/abiScanModeInfo.html
+    mode = "M6"
```
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/datalib/goes.py:203]: Thanks! I'll make some unit tests in which I pull data before 4/2/2019 as well.

File: pycontrails/datalib/goes.py [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+
+from __future__ import annotations
+
+import datetime
+import enum
+import io
+import logging
+from typing import Iterable
+
+import cartopy.crs as ccrs
```
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/datalib/goes.py]: The good news is that cartopy is easier to install now! cartopy 0.22 ships with the needed binaries.

1ec701641f24d272ab4b5d8e647d00b85fb8df09

File: pycontrails/datalib/goes.py [hunk 3]
```diff
@@ -0,0 +1,730 @@
...
+
+    Returns
+    -------
+    npt.NDArray[np.float32]
+        3d RGB array with ASH color scheme according to convention.
+
+    References
+    ----------
+    - `Ash RGB quick guide (the color space and color interpretations) <https://rammb.cira.colostate.edu/training/visit/quick_guides/GOES_Ash_RGB.pdf>`_
+    - `MIT thesis detail use for contrail detection <https://dspace.mit.edu/bitstream/handle/1721.1/124179/1144176218-MIT.pdf?22>`_
```
Comment (mlshapiro) [inline pycontrails/datalib/goes.py:663]: We should cite this in the Zotero, I'll push to this branch

Comment (mlshapiro) [reply pycontrails/datalib/goes.py:663]: kulikSatellitebasedDetectionContrails2019 (0836a5a4649f59ac29e340d25df0d9645536dea9)

**Review by mlshapiro — APPROVED** [review]
This looks really good, thanks!

File: pycontrails/datalib/goes.py [hunk 4]
```diff
@@ -0,0 +1,730 @@
...
+            da = da.expand_dims(band_id=ds["band_id"].values)
+
+        # Attach some useful attrs -- only using goes_imager_projection currently
+        da.attrs["goes_imager_projection"] = ds.goes_imager_projection.attrs
+        da.attrs["geospatial_lat_lon_extent"] = ds.geospatial_lat_lon_extent.attrs
+
+        return da
+
+
+def extract_goes_viz_artifacts(
```
Comment (mlshapiro) [inline pycontrails/datalib/goes.py:552]: I'm not keen on this name - what if it was `extract_goes_visualization` or `extract_goes_image`?

File: pycontrails/datalib/goes.py [hunk 5]
```diff
@@ -0,0 +1,730 @@
+"""Support for GOES access and analysis.
+
+Resources
+---------
+
+- `GOES on AWS notes <https://docs.opendata.aws/noaa-goes16/cics-readme.html>`_
```
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/datalib/goes.py:6]: We are! I'll add a link to the GCP info. I found the AWS notes really helpful for figuring out the naming conventions, so I'll leave that link as well.

File: pycontrails/datalib/goes.py [hunk 6]
```diff
@@ -0,0 +1,730 @@
...
+    region : GOESRegion | str = {"F", "C", "M1", "M2"}
+        GOES Region of interest. Uses the following conventions.
+
+        - F: Full Disk
+        - C: CONUS
+        - M1: Mesoscale 1
+        - M2: Mesoscale 2
+
+    channels : str | set[str] | None
+        Set of channels or bands for CMIP data. The 16 possible channels are
```
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/datalib/goes.py:293]: I think CMIP (cloud and moisture imagery product) is the correct "product type" name ... but it looks like both names are used somewhat interchangably. I'll keep as CMIP. This string actually appears in the GCS URI.

File: pycontrails/datalib/goes.py (hunk 4 above)
Comment (zebengberg) [reply pycontrails/datalib/goes.py]: 481d223b7a633578209b1f10fd5f3d5d9c8fc8b1

File: pycontrails/datalib/goes.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/datalib/goes.py:7]: 481d223b7a633578209b1f10fd5f3d5d9c8fc8b1

File: pycontrails/datalib/goes.py (hunk 3 above)
Comment (zebengberg) [reply pycontrails/datalib/goes.py]: 481d223b7a633578209b1f10fd5f3d5d9c8fc8b1

File: pycontrails/datalib/goes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/datalib/goes.py]: f290ca642471a3746f9c62f98f707d6b9b1ace28

File: pycontrails/datalib/goes.py [hunk 7]
```diff
@@ -0,0 +1,587 @@
...
+           [274.16742, 274.11523, 276.5148 , 273.85443, 270.51593],
+           [274.84555, 275.15854, 272.60248, 270.67242, 272.23734]],
+          dtype=float32)
+
+    >>> # The data has been cached locally
+    >>> assert goes.cachestore.listdir()
+
+    >>> # Download GOES data directly into memory by setting cachestore=None
+    >>> goes = GOES(region="M2", channels=("C11", "C12", "C13"), cachestore=None)
+    >>> da = goes.get("2021-04-03 02:10:00")
```
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/datalib/goes.py:360]: Thanks!!

f290ca642471a3746f9c62f98f707d6b9b1ace28

**Pull request merged**